staging/src/k8s.io/apiserver/pkg/endpoints/request/requestcontext.go

```diff
@@ -40,37 +41,55 @@ type RequestContextMapper interface {
 }
 
 type requestContextMap struct {
-	contexts map[*http.Request]Context
-	lock     sync.Mutex
+	contexts map[*http.Request]*atomic.Value
+	lock     sync.RWMutex
 }
 
 // NewRequestContextMapper returns a new RequestContextMapper.
 // The returned mapper must be added as a request filter using NewRequestContextFilter.
 func NewRequestContextMapper() RequestContextMapper {
 	return &requestContextMap{
-		contexts: make(map[*http.Request]Context),
+		contexts: make(map[*http.Request]*atomic.Value),
 	}
 }
 
+func (c *requestContextMap) getValue(req *http.Request) (*atomic.Value, bool) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	value, ok := c.contexts[req]
+	return value, ok
+}
+
+type contextWrap struct {
+	Context
+}
+
 // Get returns the context associated with the given request (if any), and true if the request has an associated context, and false if it does not.
 // Get will only return a valid context when called from inside the filter chain set up by NewRequestContextFilter()
 func (c *requestContextMap) Get(req *http.Request) (Context, bool) {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	context, ok := c.contexts[req]
-	return context, ok
+	value, ok := c.getValue(req)
+	if !ok {
+		return nil, false
+	}
+
+	if context, ok := value.Load().(*contextWrap); ok {
+		return context.Context, ok
+	}
+
+	return nil, false
 }
 
 // Update maps the request to the given context.
 // If no context was previously associated with the request, an error is returned and the context is ignored.
 func (c *requestContextMap) Update(req *http.Request, context Context) error {
-	c.lock.Lock()
-	defer c.lock.Unlock()
-	if _, ok := c.contexts[req]; !ok {
+	value, ok := c.getValue(req)
+	if !ok {
 		return errors.New("No context associated")
 	}
-	// TODO: ensure the new context is a descendant of the existing one
-	c.contexts[req] = context
+	if wrapper, ok := value.Load().(*contextWrap); ok {
+		wrapper.Context = context
```

Comment:
Why does this not race with line 76? I think it does. You store a pointer. The actual wrapper object is the same.

Reply:
one request context is only use by one groutine

Reply:
if necessary, can change this by storing contextWrap not pointer.

Reply:
Done, fixed

Reply:
> one request context is only use by one groutine

I don't think this is necessarily true, at least for reading.

Reply:
And if that was the case, we would not need the atomic value.

Reply:
Yes, fixed now